src/research_index_backend/doi.py

```diff
@@ -0,0 +1,195 @@
+    ) -> None:
+
+        if not list_of_dois:
+            raise ValueError("DOI list cannot be empty")
+        if (limit <= 0) or (limit > len(list_of_dois)):
+            raise ValueError(
+                "Limit must be positive and less than the number of DOIs"
+            )
+
+        self.list_of_dois = [
+            doi.strip()
+            .rstrip(".")
+            .replace("https://doi.org/", "")
+            .replace("doi.org/", "")
+            for doi in list_of_dois
+        ]
+        self.limit = limit
+        self.update_metadata = update_metadata
+        self.doi_tracker: DOITracker = {
+            doi: DOI(doi=doi) for doi in self.list_of_dois[: self.limit]
+        }
+        self.PATTERN = compile(DOI_PATTERN, IGNORECASE)
+
+    def start_ingestion(self):
+        self.start_time = time.time()
+
+    def end_ingestion(self):
+        self.end_time = time.time()
+
+    def pattern_check(self):
+        try:
+            self.valid_pattern_dois = []
+            self.invalid_pattern_dois = []
+
+            for doi in self.doi_tracker:
+                if search := self.PATTERN.search(doi):
+                    logger.debug(f"Valid DOI pattern: {search.group()}")
+                    self.doi_tracker[doi].valid_pattern = True
+                    self.valid_pattern_dois.append(doi)
+                else:
+                    logger.warning(f"Invalid DOI pattern: {doi}")
+                    self.invalid_pattern_dois.append(doi)
+            self.num_valid_pattern_dois = len(self.valid_pattern_dois)
+            self.num_invalid_pattern_dois = len(self.invalid_pattern_dois)
+        except Exception as e:
+            logger.error(f"Error whilst checking DOI pattern: {e}")
+            raise
+
+    @connect_to_db
+    def search_dois(self, db: Driver):
+        if not self.valid_pattern_dois:
+            msg = "No DOIs have passed the pattern check and make sure to run pattern check first."
+            logger.warning(msg)
+            raise ValueError(msg)
+        query = """
+            UNWIND $dois as doi
+            OPTIONAL MATCH (o:Output {doi: doi})
+            RETURN doi, COUNT(o) > 0 as exists"""
+        try:
+            results, _, _ = db.execute_query(
+                query, dois=self.valid_pattern_dois
+            )
+            self.existing_dois = [
+                record["doi"] for record in results if record["exists"]
+            ]
+            self.new_dois = [
+                record["doi"] for record in results if not record["exists"]
+            ]
+            for doi in self.doi_tracker:
+                if doi in self.existing_dois:
+                    self.doi_tracker[doi].already_exists = True
+
+            self.num_new_dois = len(self.new_dois)
+            self.num_existing_new_dois = len(self.existing_dois)
+
+        except Exception as e:
+            logger.error(f"Error whilst searching for DOIs: {e}")
+            raise
```

Comment:
raise a specific error e.g. RunTimeError?

Reply:
Or, reduce the scope of the `try` by using the `try`,`except`, `else`, `finally` pattern?